base/applications/regedit/find.c

```diff
@@ -91,11 +91,18 @@ static BOOL
 CompareData(
     DWORD   dwType,
     LPCWSTR psz1,
-    LPCWSTR psz2)
+    size_t  cb1,
+    LPCWSTR psz2,
+    size_t  cb2)
 {
-    INT i, cch1 = wcslen(psz1), cch2 = wcslen(psz2);
+    ssize_t i;
+    size_t cch1, cch2;
```

Comment:
Could you please consider adding this statement here. These types of keys should be excluded.

   if (dwType == REG_BINARY) return FALSE;

Reply:
Rejected. Our regedit should search the binary data by string.

Reply:
Do you know if W2K3 can handle this data in its search?

Reply:
![2k3canfind](https://user-images.githubusercontent.com/2107452/224480131-64b9c277-f082-4e97-98ca-64586beeaad9.png)
Win2k3 can find the binary data by string.

Reply:
Since you are using by definition LPCWSTR for psz1 and psz2, there will not be a case of these types of variables where the BYTE count will be odd. I have seen the byte count from a REG_BINARY data be odd. If you intend to handle the REG_BINARY type, then I believe that you will need quite a bit more code. You may want to make this another PR?

Reply:
![VirtualBox_Windows Server 2003_11_03_2023_19_55_05](https://user-images.githubusercontent.com/2107452/224480346-4871121d-2f19-4c52-b705-4bb1d3fd1bd5.png)

Win2k3 **cannot** find the binary data by string.

Reply:
![unicode](https://user-images.githubusercontent.com/2107452/224480476-cb9af569-03f8-4f39-82f2-3b340a1e05a6.png)
Win2k3 *cannot* search Unicode text of the binary data.

Reply:
Okay, your suggestion is accepted. We prohibit to search in binary data by string.

Reply:
`(dwType == REG_SZ || dwType == REG_EXPAND_SZ)`
is enough condition.

Reply:
> definition LPCWSTR for psz1 and psz2

`CompareData` parameters are not guaranteed to follow the definition of `LPCWSTR`.

Reply:
Our registry can be broken, so we need strict checking...

Reply:
> I believe that you will need quite a bit more code

Now simplified by deleting data size paramters.

Reply:
It is perfectly valid for `RegQueryValue` to return a non-null terminated string.

Reply:
> `(dwType == REG_SZ || dwType == REG_EXPAND_SZ)` is enough condition.

And what about `REG_MULTI_SZ` ?

Reply:
I believe `RegQueryValueExW` will null-terminate the string for you, even if it's stored without a null-terminator in the registry. So checking for type seems reasonably safe.
Multi-SZ and binary would need special handling if we want to support them, since wcsstr won't do the trick. I'm in favor of supporting Multi-SZ, but wouldn't make it a requirement here, since the functionality is already broken anyway.

Reply:
https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regqueryvalueexw
> To ensure that any string values (REG_SZ, REG_MULTI_SZ, and REG_EXPAND_SZ) returned are null-terminated, use the [RegGetValue](https://learn.microsoft.com/en-us/windows/desktop/api/winreg/nf-winreg-reggetvaluea) function.

However the RegGetValue function is Vista+.

Reply:
The RegQueryValueExW is not without its problems though. See
https://devblogs.microsoft.com/oldnewthing/20040824-00/?p=38063

Reply:
Regarding RegGetValue I see "Minimum supported server | Windows Server 2008, Windows Server 2003 with SP1"

Reply:
Huh I guess I misremembered. RegQueryValueExW is still fine -- just either null-terminate the buffer yourself or pass the size to all callees and use strsafe functions.